go/src/koding/klient/machine/mount/index/index.go

```diff
@@ -0,0 +1,328 @@
+package index
+
+import (
+	"crypto/sha1"
+	"encoding/json"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+)
+
+// Entry represents a single file registered to index.
+type Entry struct {
+	Name  string      `json:"name"`  // The relative name of the file.
+	CTime int64       `json:"ctime"` // Metadata change time since EPOCH.
+	MTime int64       `json:"mtime"` // File data change time since EPOCH.
+	Mode  os.FileMode `json:"mode"`  // File mode and permission bits.
+	Size  int64       `json:"size"`  // Size of the file.
+	SHA1  []byte      `json:"sha1"`  // SHA-1 hash of file content.
+}
+
+// NewEntryFile creates new Entry from a file stored under path argument.
+// Info is optional and, if given, should store LStat result on the given file.
+func NewEntryFile(root, path string, info os.FileInfo) (e *Entry, err error) {
+	if info == nil {
+		if info, err = os.Lstat(path); err != nil {
+			return nil, err
+		}
+	}
+
+	// Get relative file name.
+	name, err := filepath.Rel(root, path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Compute file's SHA-1 sum.
+	var sum []byte
+	if !info.IsDir() {
+		if sum, err = readSHA1(path); err != nil {
```

Comment:
you are calculating the hash just from the file content, git hashing uses other kind of info ( commit message, author, committer, previous hash etc ), how are you planning to keep them insync? or from another angle, git has history of commit orders, this implementation of index does not have any kind of ordering/sorting. Another point for git indexes would be, files could be in the folder (uncommitted) without a hash value, but in any case you need to keep them in the index for various reasons.

I dont have a strong opinion at this moment since we will see what is waiting for us in the future, i just wanted to give you a speed bump of 10x :)